test(negative): implement backupstore setup

yangchiu opened the pull request:
test: implement backupstore setup

(1) add backupstore.py for robot test cases
(2) only support s3 now, the subprocess parts in backupstore.py need to be refined to make nfs work
(3) fix wrong longorn client url issue when using it out-of-cluster (the same issue as https://github.com/longhorn/longhorn-tests/pull/1300#discussion_r1273250265)

For https://github.com/longhorn/longhorn/issues/6710

Signed-off-by: Yang Chiu <yang.chiu@suse.com>

innobead reviewed (changes requested):
Some feedbacks, but I didn't review all of them. Some needs to resolve first.

backupstore.py seems verbose and unstructured. I would suggest doing below.

Use different python modules to aggregate the related codes.
- backupstore/__init__.py including all common functions for backupstore
   backupstore/nfs.py, minio.py
 
 Also, using the common naming (verb) for the function and method name instead, so `backupstore_cleanup` should be changed to `cleanup_backupstore` and add it to backupstore/__init__.py. 
 
 Please review the code and do corresponding update.

yangchiu commented on e2e/libs/backupstore.py:
```diff
@@ -0,0 +1,736 @@
+TEMP_FILE_PATH = "/tmp/temp_file"
+
+RETRY_COUNT = 300
+RETRY_INTERVAL = 1
```
This can be refined later if it actually has the situation that deleting backup volumes is unexpected time-consuming. If it's observed, it can be refined.

yangchiu commented on e2e/libs/backupstore.py:
```diff
@@ -0,0 +1,736 @@
+
+SETTING_BACKUP_TARGET = "backup-target"
+SETTING_BACKUP_TARGET_CREDENTIAL_SECRET = "backup-target-credential-secret"
+SETTING_BACKUPSTORE_POLL_INTERVAL = "backupstore-poll-interval"
```
Moved these backup specific values to a backup class which contains all backup specific code.

yangchiu commented on e2e/libs/backupstore.py:
```diff
@@ -0,0 +1,736 @@
+            backupstores[i] = backupstores[i].split(":")[0]
+    except ValueError:
+        backupstores = backupstore.split(":")[0]
+    return backupstores
```
Removed.

yangchiu commented on e2e/libs/backupstore.py:
```diff
@@ -0,0 +1,736 @@
+    client.update(backup_target_credential_setting, value="")
+    backup_store_poll_interval = client.by_id_setting(
+        SETTING_BACKUPSTORE_POLL_INTERVAL)
+    client.update(backup_store_poll_interval, value="300")
```
Refined.

yangchiu commented on e2e/libs/backupstore.py:
```diff
@@ -0,0 +1,736 @@
+        umount_nfs_backupstore(client)
+
+
+def minio_get_api_client(client, core_api, minio_secret_name):
```
Refined.

yangchiu commented on e2e/libs/backupstore.py:
```diff
@@ -0,0 +1,736 @@
+                 secure=True)
+
+
+def minio_get_backupstore_bucket_name(client):
```
Refined.

yangchiu commented:
> Some feedback, but I didn't review all of them. Some need to be resolved first.
> 
> backupstore.py seems verbose and unstructured. I would suggest doing the below.
> 
> Use different Python modules to aggregate the related codes.
> 
>     * backupstore/__init__.py including all common functions for backupstore
>       backupstore/nfs.py
>       backupstore/minio.py (but actually it's s3)
> 
> 
> Also, using the common naming (verb) for the function and method name instead, `backupstore_cleanup` should be changed to `cleanup_backupstore` and added to backupstore/__init__.py.
> 
> Please review the code and do the corresponding update.

Refined to have class structure.

yangchiu commented:
@longhorn/qa @innobead This PR is required to run negative test case `Reboot Volume Node While Heavy Writing And Recurring Jobs Exist` in v1.4.4 release test plan.

c3y1huang commented on e2e/libs/backupstore/base.py:
```diff
@@ -0,0 +1,198 @@
+            time.sleep(RETRY_INTERVAL)
+        assert not found
+
+    def cleanup_backupstore(self, client):
```
```suggestion
    def cleanup_backup_volumes(self, client):
```

yangchiu replied:
Refined.

c3y1huang commented on e2e/libs/backupstore/minio.py:
```diff
@@ -0,0 +1,205 @@
+        if len(data) == 0:
+            data = {"testkey": "test data from mino_create_file_in_backupstore()"}
+
+        with open(TEMP_FILE_PATH, 'w') as temp_file:
```
Where is TEMP_FILE_PATH defined?

yangchiu replied:
Refined.

yangchiu commented on e2e/libs/backupstore/base.py:
```diff
@@ -0,0 +1,198 @@
+    def set_backupstore(self, client):
+        return NotImplemented
+
+    def reset_backupstore_setting(self, client):
```
Can be done later when we have setting related test cases since currently backupstore setting is the only setting exists in the test code.

yangchiu replied:
Sure. I'll organize them in the later PR.

c3y1huang replied:
Given that we are introducing this new function, I don't see any reason we shouldn't organize this now rather than deferring to a later PR.

yangchiu replied:
Refined. Thank you!

yangchiu commented on e2e/libs/backupstore/minio.py:
```diff
@@ -0,0 +1,205 @@
+from minio.error import ResponseError
+from urllib.parse import urlparse
+
+class Minio(Base):
```
Not sure about this. Even though the class name is changed to `S3`, the apis called in this class are still from `minio`.

yangchiu commented on e2e/keywords/common.resource:
```diff
@@ -21,6 +22,7 @@ Set test environment
     Set Test Variable    ${deployment_list}
     @{statefulset_list} =    Create List
     Set Test Variable    ${statefulset_list}
+    set_backupstore
```
Put it here to implicitly setup backupstore.

c3y1huang reviewed (dismissed):
I assume most of the codes are moved from the integration tests as I see there are some variables that are still undefined, and the functions could be better organized. However, as long as the tests can run I think it's fine to merge this first and enhance later.

cc @innobead

yangchiu commented:
Need to refine:
1. Rebooting volume node will cause the volume fast detaching and reattaching again **at the moment of the node back to normal**, currently the test case is unable to catch the timing and causes assertion failed.
2. If the node of minio service is rebooted, it's unable to create backup too, and causes the test case failed.

yangchiu commented:
>     1. Rebooting volume node will cause the volume fast detaching and reattaching again **at the moment of the node back to normal**, currently the test case is unable to catch the timing and causes assertion failed.

Fixed by rearrange the orders of status checking logic.

>     2. If the node of minio service is rebooted, it's unable to create backup too, and causes the test case failed.

Fixed by creating a deployment for minio instead of a plain pod.

And  according to https://github.com/longhorn/longhorn/issues/7956:

> Pushing a backup and a purge once a minute may be a little aggressive.
> Maybe the test can be slightly modified to recur less often.

Refined the recurring job schedule to every 2 minutes.

yangchiu commented:
@innobead Is it possible to merge this PR? Since the test cases have been listed in release testing plans and they are always blocked by this PR.

innobead commented:
@yangchiu sounds good. @c3y1huang Please help review this. After that, we can merge this PR.

yangchiu commented:
@longhorn/qa Please help review this. Many thanks.

c3y1huang commented on e2e/libs/backupstore/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from backupstore.nfs import Nfs
+from backupstore.minio import Minio
```
GitHub is showing a red marker (No newline at end of file) for some files in the PR. This is probably not harmful, but it can be rather bothersome. Suggest adding a new line at the end of those files to resolve it.

yangchiu replied:
Refined. Thanks!

yangchiu commented on e2e/libs/recurringjob/recurringjob.py:
```diff
@@ -19,7 +19,7 @@ def create(self,
                task="snapshot",
                groups=[],
-               cron="* * * * *",
+               cron="*/2 * * * *",
```
Please refer to https://github.com/longhorn/longhorn-tests/pull/1539#issuecomment-1951537233

yangchiu replied:
Refined. Thanks!

yangchiu commented on e2e/libs/recurringjob/rest.py:
```diff
@@ -45,144 +47,142 @@ def add_to_volume(self, job_name, volume_name):
     def cleanup(self, volume_names):
         for volume_name in volume_names:
-            logging(f"Cleaning up recurringjobs for volume {volume_name}")
+            logging(f"Cleaning up recurring jobs for volume {volume_name}")
```
One thing needs to be mentioned is longhorn client itself uses naming like create_recurring_job‎ and by_id_recurring_job, even though this changes to single word, there is still inconsistency in the file.

c3y1huang replied:
I think that it's fine since it's external to the longhorn-test repo. I understand there are some inconsistencies in the longhorn REST client. Let's just focus on the e2e test codebase now.

yangchiu replied:
Refined. Thanks!

yangchiu commented on e2e/libs/backupstore/base.py:
```diff
@@ -0,0 +1,198 @@
+import time
+import hashlib
+
+SETTING_BACKUP_TARGET = "backup-target"
+SETTING_BACKUP_TARGET_CREDENTIAL_SECRET = "backup-target-credential-secret"
+SETTING_BACKUPSTORE_POLL_INTERVAL = "backupstore-poll-interval"
+
+BACKUPSTORE_BV_PREFIX = "/backupstore/volumes/"
+BACKUPSTORE_LOCK_DURATION = 150
+
+RETRY_COUNT = 300
+RETRY_INTERVAL = 1
```
I'm still not sure about this. WDYT? @longhorn/qa

c3y1huang replied:
As the test grows, it often accumulates numerous constants, like [this](https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/common.py#L30-L302). I believe it would be more organized to manage these contents is dedicated 
 constant file.
 
However, if the team decides not to take this approach, I will understand that decision. As such, we then need to take out the constants in the constant.py files and place them in the relevant places.

yangchiu replied:
It's because `common.py` is written in C-like style and mixing everything together. We can use a more OOP way to handle them. And I think it is reasonable for constants placed in different files if they are specific to and only used by a particular class.

c3y1huang replied:
I personally prefer the constants in a separate file for easier management, but I am flexible if the team decides otherwise. I believe what's most important is making sure that @longhorn/qa agrees and understands the final decision and that we maintain a consistent convention in the e2e.

roger-ryao replied:
Based on my previous experience, my habit is similar to @c3y1huang's suggested usage approach, where I centralize the use of variables in the same variable file, `constants.py`. In the future, when the data in the variable file becomes extensive, it will be easier for us to organize and categorize them further.
However, I currently don't have a concept of the method @yangchiu mentioned: `We can use a more OOP way to handle them.`

yangchiu replied:
Refined. Thank you!

yangchiu commented on e2e/libs/keywords/backupstore_keywords.py:
```diff
@@ -0,0 +1,55 @@
+    def delete_dummy_in_progress_backup(self, volume_name):
+        client = get_longhorn_client()
+        core_api = client.CoreV1Api()
+        delete_backup_cfg_file(client,
```
Refined. Thanks!

c3y1huang commented on e2e/libs/keywords/backupstore_keywords.py:
```diff
@@ -0,0 +1,55 @@
+                                               core_api,
+                                               volume_name,
+                                               backup_name,
+                                               corrupt_backup_cfg_data)
```
Add line break at EOF. Please also check other files.

yangchiu replied:
Refined. Thank you!

c3y1huang commented on e2e/libs/backupstore/base.py:
```diff
@@ -0,0 +1,134 @@
+        return NotImplemented
+
+    def get_backup_target(self, client):
+        backup_target_setting = client.by_id_setting(SETTING_BACKUP_TARGET)
```
`SETTING_BACKUP_TARGET` is not defined. Please check other places.

yangchiu replied:
Refined. Thank you!

c3y1huang commented on e2e/libs/backupstore/minio.py:
```diff
@@ -0,0 +1,185 @@
+        base64_minio_access_key = secret.data['AWS_ACCESS_KEY_ID']
+        base64_minio_secret_key = secret.data['AWS_SECRET_ACCESS_KEY']
+        base64_minio_endpoint_url = secret.data['AWS_ENDPOINTS']
+        base64_minio_cert = secret.data['AWS_CERT']
```
I couldn't find where this is used. Consider Removing if not required.

yangchiu replied:
Removed. Thank you!

c3y1huang commented on e2e/libs/recurringjob/rest.py:
```diff
@@ -45,139 +47,137 @@ def add_to_volume(self, job_name, volume_name):
         current_time = datetime.utcnow()
         current_timestamp = current_time.timestamp()
-
+        logging(f"Recorded current time = {current_time}, timestamp = {current_timestamp}")
```
Could you explain the purpose of this log?

yangchiu replied:
Removed. Thank you!

c3y1huang commented on e2e/libs/utility/utility.py:
```diff
@@ -187,3 +187,12 @@ def get_name_suffix(*args):
     return suffix
+
+
+def get_backupstores():
```
This looks like it belongs in the backupstore lib.

yangchiu replied:
Refined. Thank you!

yangchiu commented on e2e/libs/backupstore/minio.py:
```diff
@@ -0,0 +1,184 @@
+
+    def create_file_in_backupstore(self, client, core_api, file_path, data={}): # NOQA
+        backup_target_credential_setting = client.by_id_setting(
+            SETTING_BACKUP_TARGET_CREDENTIAL_SECRET)
```
Refined. Thank you!
